### ISSUE TITLE
[BUG][mthreads] 标准模式 c64 高并发 decode 吞吐塌陷——根因=sqlite autotune 缓存全量反射（automap.prepare 3s/1426表）

### ISSUE THREAD

**Issue by JosephNew** [flagos-ai/FlagGems#6390]
## 现象

摩尔 MTT S5000 标准模式（tp=4，Qwen3.6-27B），c64 高并发 bench 压测下 decode 吞吐塌陷到 ~3 tok/s，GPU 利用率几乎为 0，服务不崩但性能不可用。

## 观测数据（两轮回归一致）

| 指标 | rc0 c8 参考 | rc2 tag 版 c64 | rc2 HEAD c64 |
| --- | --- | --- | --- |
| decode 吞吐 | 40.19 tok/s | 3.10 tok/s | 1.80~3.39 tok/s |
| GPU 利用率 | — | 3-8% | 0% |
| scheduler_TP0-3 CPU | — | 99% | 99% |
| bench 进度 | c8 秒级 | 34min 6-7/64 | 1h 9/64（ETA 9h+） |

- serve 日志 `gen throughput (token/s): 3.10`，两条 Decode batch 日志间隔 14 分钟。
- `mthreads-gmi` 显示卡 GPU 利用率 0-8%，但 `scheduler_TP0-3` 进程 CPU 99%。

## 环境

- 机器：`10.121.38.26`（8 卡 S5000 80GB），卡 4,5,6,7。
- 版本：FlagGems rc2 分支 HEAD `77684a9` / FlagTree `84b74ff`（含 #1103）/ plugin `1c84db7`。
- 标准模式 env：`SGLANG_FL_PER_OP=topk=reference` + flagcx + `SGLANG_MUSA_FP32_TP_ALLREDUCE=1`。
- bench：`sglang.bench_serving` random-ids 1024/1024，max-concurrency 64，num-prompts 64。

## 复现

起服务（标准模式，tp=4）后：
```bash
source /root/.virtualenvs/sglang-0.5.6/bin/activate
export PYTHONPATH=/sgl-workspace/sglang/python
python3 -m sglang.bench_serving --backend sglang --host 127.0.0.1 --port 30000 --model /datapool/models/Qwen3.6-27B --tokenizer /datapool/models/Qwen3.6-27B --dataset-name random-ids --random-input-len 1024 --random-output-len 1024 --max-concurrency 64 --num-prompts 64
```

## 初步判断（host-bound，待开发定位）

**CPU 99% + GPU 0-8%** 是典型的 host-bound：不是 GPU 算力瓶颈，而是 CPU 端调度/算子开销吃满。可疑方向：

1. **flag_gems 算子 host 开销**：标准模式 decode 每 step 大量小 shape 算子走 flag_gems dispatch，若 host 端 Python 开销（dispatch / autotune 查询 / shape 处理）随并发放大，会打满 scheduler CPU。
2. **flagcx 高并发通信 host 开销**。
3. **sglang scheduler 高并发特性**（但 nvidia/海光无此严重塌陷，倾向摩尔特有）。

## 排除项

- **非 #1173 / flagtune 死锁**：该问题已随 PR #1103 修复（capture 退化 + 首请求死锁均解除），吞吐塌陷在修复后依然存在。
- **非 #6121 sqlite 锁死**：WAL 修复有效，c64 压 1h 0 次锁死，服务不崩。

## 诉求

请开发定位 c64 高并发下 host-bound 的根因（是 flag_gems 算子 host 开销、flagcx、还是 sglang scheduler），并给出是否可通过 env / 黑名单绕过的方案。目前 c64 场景完全不可用，需逐级（c8/c16/c32/c64）定位退化点。

**Comment by JosephNew** [comment]
## 根因已定位（host-bound 的精确来源：sqlite autotune 缓存全量反射）

用 py-spy dump `scheduler_TP0-3` 进程（CPU 99% 时）+ 实测 sqlite 库，锁定了吞吐塌陷的根因链。

### 现象回顾

c64 高并发下 decode 吞吐 ~3 tok/s，GPU 利用率 0%，scheduler 进程 CPU 99%（4 个 TP worker 都 99%）。

### 根因链（证据完整）

**① 元凶：`SQLPersistantModel.build_sql_model_by_db` 每次 config-cache miss 都全量反射整个 sqlite 库**

`flag_gems/utils/models/sql.py:164-171`：
```python
@staticmethod
def build_sql_model_by_db(name, engine):
    AutoBase = sqlalchemy.ext.automap.automap_base()
    AutoBase.prepare(engine)   # ← 反射整个 db 的所有表 + 索引
    ModelCls = AutoBase.classes.get(name)
    return ModelCls
```

`automap.prepare(engine)` 是**全量反射**——为库里每一张表建 mapper、配置列、反射索引，与「这次只想查一张表」完全无关。

**实测**（摩尔 S5000 容器内，直接连 db 计时）：
```
automap prepare 全量反射耗时: 3.08 s, 表数: 1426
```

一次反射 = 3 秒纯 host CPU 活，GPU 完全空转。

**② 放大器：decode 阶段 index 算子的 shape 随 batch size 变化 → config-cache 频繁 miss**

py-spy 栈（反复出现）：
```
build_sql_model_by_db (sql.py:171)          ← 全量反射
  get_sql_model (sql.py:237)
    get_config (sql.py:273)
      __contains__ (libentry.py:201)        ← config_key not in cache
        run (libentry.py:1015)
          _index_linearized_wrapper (index_linearized_..._m_32_n_5120.py:70)
            _run_linearized_adjacent_index (index.py:686)
              index (index.py:800)
                _get_pruned_states (logits_processor.py:447)
                  forward (logits_processor.py:322)
```

decode 阶段 `logits_processor._get_pruned_states`（speculative decode pruning）走 flag_gems 的 `index` 算子，其 shape `m`（batch size）在 c64 下从 1 到 64 变化。每个新 shape → 新 config key → 新 `short_name` → `sql_model_pool` miss → 触发一次全量反射。

**③ 恶性循环：反射完建新表 → 表数膨胀 → 反射越来越慢**

sqlite 库现状：
```
total tables: 1409（index 相关 1382 张）
```

每个新 shape 建一张新表（`_index_linearized_jit_function_<hash>-<hash>`），表数持续膨胀，每次全量反射的耗时随之线性增长。这解释了 bench「越跑越慢」：14/64 时 ETA 5h+，20/64 时 ETA 2h+。

### 为什么 c1 快 / c64 慢

- c1：batch size 固定，config key 极少，反射触发次数少
- c64：batch size 1..64 全变，config key 数量 ~64 倍，每个首次出现都触发一次 3s+ 全量反射

### 已做验证（排除其他假设）

1. ✅ py-spy 栈稳定复现：热点恒在 `build_sql_model_by_db` → `automap.prepare` → `get_indexes`/`map`/`_configure_property`
2. ✅ 实测全量反射 3.08s / 1426 表
3. ✅ ConfigCache 加进程内内存缓存**无效**（本地打补丁验证）——因为瓶颈不是「同一 key 重复查」，而是「每个新 shape 首次查询都触发全量反射」，内存缓存救不了首次 miss 的反射开销

### 修复建议

**治本（改动小、收益大）**：`build_sql_model_by_db` 不要再 automap 全量反射。两个方向：
1. 用 `sqlalchemy.inspect(engine).has_table(short_name)`（O(1)）判断表存在性，替代 automap 全量反射；或
2. 直接去掉 `build_sql_model_by_db` 这条路径——`build_sql_model_by_py` + `CreateTable(if_not_exists=True)` 已经能安全兜底建表，根本不需要先反射 db 查表。

**放大器（可选）**：decode 阶段 `logits_processor._get_pruned_states` 的 index 算子 shape 随 batch 变化（本应在 cuda graph 内 shape 固定）。若该 index 算子能进 cuda graph 或 shape 稳定，config key 数量会大幅下降，进一步降低 sqlite 压力。

> 这个根因与 #6121（sqlite 并发写锁死）同属 flag_gems autotune 持久化缓存家族——#6121 修了写路径（WAL+busy_timeout），本次是读路径（全量反射）的性能问题。建议统一审视 sqlite 缓存的热路径开销。

**Comment by JosephNew** [comment]
## 已定位 + 修复 + 提 PR（#6416）

根因已修复并本地真机验证通过，PR：**https://github.com/flagos-ai/FlagGems/pull/6416**（base=5.4.0-rc2）。

### 修复方案

`build_sql_model_by_db` 从「automap 全库反射」改为「`inspect(engine).has_table()` O(1) 判断 + 单表 `Table(autoload_with=engine)` 反射 + `automap_base(metadata=metadata)`」。语义不变（反射列完全一致），反射范围从全库收窄到单表。

### 真机验证数据（mthreads S5000）

| 指标 | 修复前 | 修复后 |
| --- | --- | --- |
| 单表反射 vs 全库反射 | 5.108s（1611 表） | 0.005s（~1000×） |
| c16 decode 吞吐 | 6.10 tok/s（TPOT 2018ms） | **47.83 tok/s**（TPOT 145.67ms） |
| c32 decode 吞吐 | 跑不完 | **95.09 tok/s**（TPOT 182.61ms） |
| cuda graph capture | 56s | **14.6s** |

py-spy 确认热点消失：`automap.prepare` → `get_indexes` 不再出现，scheduler 回到正常 `recv_requests`。

### 附带收益

不只摩尔线程——任何让 autotune 缓存表数膨胀到百/千级的厂商/场景都会被全量反射拖慢，此修复通用。

**Comment by JosephNew** [comment]
## 修复已合入，且与本 issue 中的验证版本**逐字一致** → 无需复测，收口

PR **#6416** 已于 **2026-09-18 02:30** 合入 `5.4.0-rc2`（squash 为 commit `ad89ec477f7c0533176d791bb9161809191315c6`，父提交 `55114639`）。

### 为什么不再复测

对比「PR 分支上的原始提交」（`51dfed75`）与「合入后的 commit」（`ad89ec47`）：**补丁正文逐字相同**，`+10/-2`，改动文件同为 `src/flag_gems/utils/models/sql.py`，唯一差异是 hunk 行号偏移（`@@ -165,10 +165,18 @@` → `@@ -176,10 +176,18 @@`）——因为合入基线在 PR 打开后还前进了 11 行（同文件上的 #6392，SQLAlchemy 1.4 兼容，02:13 合入）。

可自行复核（去掉 `@@` 行号头后逐字比对，输出 `逐字一致`）：

```bash
gh api repos/flagos-ai/FlagGems/commits/51dfed75 --jq '.files[].patch' | grep -v '^@@' > /tmp/pr_head.patch
gh api repos/flagos-ai/FlagGems/commits/ad89ec47 --jq '.files[].patch' | grep -v '^@@' > /tmp/merged.patch
diff /tmp/pr_head.patch /tmp/merged.patch && echo "逐字一致"
```

两份 patch 的 hunk 头分别为 `@@ -176,10 +176,18 @@`（合入版）与 `@@ -165,10 +165,18 @@`（PR 原始版）。

维护者未对该修复做任何调整，合入码 = 我提交 PR 前已真机验证过的码。因此下表的验证数据即为该修复在目标版本上的最终数据，无需再跑一轮。

### 修复内容（最终合入形态，rc2 现状）

`src/flag_gems/utils/models/sql.py:187+`：

```python
if not sqlalchemy.inspect(engine).has_table(name):
    return None
metadata = sqlalchemy.MetaData()
sqlalchemy.Table(name, metadata, autoload_with=engine)
AutoBase = sqlalchemy.ext.automap.automap_base(metadata=metadata)
AutoBase.prepare()
ModelCls = AutoBase.classes.get(name)
return ModelCls
```

即：`inspect(engine).has_table()` O(1) 判存在 + 单表 `Table(autoload_with=engine)` 反射 + `automap_base(metadata=metadata)`，替换原先的 `automap_base().prepare(engine)` 全库反射。语义不变（反射列完全一致）。

### 验证数据（mthreads S5000，tp=4，Qwen3.6-27B）

| 指标 | 修复前 | 修复后 |
| --- | --- | --- |
| 单表反射 vs 全库反射 | 5.108s（1611 张表） | **0.005s**（~1000×） |
| c16 decode 吞吐 | 6.10 tok/s（TPOT 2018ms） | **47.83 tok/s**（TPOT 145.67ms） |
| c32 decode 吞吐 | 跑不完 | **95.09 tok/s**（TPOT 182.61ms） |
| cuda graph capture | 56s | **14.6s** |

py-spy 复核：`automap.prepare` → `get_indexes` 热点消失，scheduler 回到正常 `recv_requests`。c64 下 GPU 利用率由 0% 恢复。

> 附注：本修复的收益不限于 mthreads——任何让 autotune 缓存表数膨胀到百/千级的场景都会被全量反射拖慢。与 #6121（sqlite 并发写锁死）同属 autotune 持久化缓存家族，#6121 修的是写路径、本 issue 是读路径。

关闭本 issue。如后续在 rc2 之后的版本上再观察到 c64 吞吐塌陷，欢迎重新打开或另开。
